src/net/client/tsig.rs

```diff
@@ -0,0 +1,704 @@
+//! A TSIG signing & verifying passthrough transport.
+//!
+//! This module provides a transport that wraps the [high-level support for
+//! signing message exchanges with TSIG][crate::tsig].
+//!
+//! # Usage
+//!
+//! 1. Create a signing [Key].
+//! 2. Create a [Connection] that wraps an upstream connection and uses the
+//!    key.
+//! 3. [Send a request][Connection::send_request] using the connection.
+//! 4. [Receive the response][Request::get_response] or responses.
+//!
+//! # How it works
+//!
+//! Supplying the key is optional. The transport only affects the request and
+//! response if a key is supplied. This allows for optional signing without
+//! having to construct a different client stack.
+//!
+//! When a key is supplied, requests are automatically signed and response
+//! signatures are automatically verified. On verification failure
+//! [Error::ValidationError][crate::net::client::request::Error] will be
+//! returned.
+#![cfg(all(feature = "tsig", feature = "unstable-client-transport"))]
+#![warn(missing_docs)]
+#![warn(clippy::missing_docs_in_private_items)]
+
+use core::ops::DerefMut;
+
+use std::boxed::Box;
+use std::fmt::{Debug, Formatter};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::vec::Vec;
+
+use bytes::Bytes;
+use tracing::{debug, trace, warn};
+
+use crate::base::message::CopyRecordsError;
+use crate::base::message_builder::AdditionalBuilder;
+use crate::base::wire::Composer;
+use crate::base::Message;
+use crate::base::StaticCompressor;
+use crate::net::client::request::{
+    ComposeRequest, ComposeRequestMulti, Error, GetResponse,
+    GetResponseMulti, SendRequest, SendRequestMulti,
+};
+use crate::rdata::tsig::Time48;
+use crate::tsig::{ClientSequence, ClientTransaction, Key};
+
+/// TODO
+#[derive(Clone, Debug)]
+enum TsigClient<K> {
+    /// TODO
+    Transaction(ClientTransaction<K>),
+    // TODO
+    //Sequence(ClientSequence<K>),
+}
+
+/// TODO
+#[derive(Clone, Debug)]
+enum TsigClientMulti<K> {
+    /// TODO
+    //Transaction(ClientTransaction<K>),
+
+    /// TODO
+    Sequence(ClientSequence<K>),
+}
+
+//------------ Connection -----------------------------------------------------
+
+#[derive(Clone)]
+/// TODO
+pub struct Connection<Upstream, K> {
+    /// Upstream transport to use for requests.
+    ///
+    /// This should be the final transport, there should be no further
+    /// modification to the request before it is sent to the recipient.
+    upstream: Arc<Upstream>,
+
+    /// TODO
+    key: Option<K>,
```

Comment:
It seems that if key is equal to None, this code silently disables TSIG processing. For a security function that is an accident waiting to happen.

Reply:
It was, if I recall correctly, envisaged as a way to use the same client instance for both signed and unsigned requests, in particular because signing is being added primarily for XFR requests and responses, but a client requiring a key would not then be usable for other types of request, which perhaps on a TCP stream it would be of interest to reuse the stream rather than have to use a separate client and a separate TCP stream.

However, I see your point, if one creates a TSIG client that is because you want to sign requests, and it is a footgun to be able to NOT sign requests with a signing client. I have removed the Option, with the assumption that users can create a separate non-signing client if needed. If the efficiency angle I just referred to about TCP streams is a point of interest we can revisit this, but for now it resolves your concern.

Reply:
You can see the effect this has by looking at https://github.com/NLnetLabs/domain/pull/335/commits/dbf80ced2dbfa376a2a90304091e01e28f197c8c. This issue occurs because the test code doesn't know in advance if it has to create a signing client or not, so has to if else branch and box the created client to erase the different types in order to use them without caring if the client is signing or not. For now I'm hoping that this is an artefact of test code and that real application code wouldn't be as affected.

Reply:
**Update:** I see now that it was even documented why the key is optional:

> Supplying the key is optional. The transport only affects the request and response if a key is supplied. This allows for optional signing without having to construct a different client stack.